### PR TITLE
chore(schema): Example for adding `log_namespace` support to `demo_logs` source

### PR DESCRIPTION
**Pull request by fuchsnj**
This PR adds `log_namespace` support to the `demo-logs` source, and serves to act as an example for what is required.
For a more complicated example, you can look at the `datadog-agent` source, which already supports log namespacing.

For reference, the Log Namespacing RFC can be viewed [here](https://github.com/vectordotdev/vector/blob/edacee2482be1e79e10ea9a3d6766114cae44e67/rfcs/2022-04-20-12187-log-namespacing.md)

Changes Required
1. Add `log_namespace: Option<bool>` to the source config. This is the source level configuration for enabling the `log_namespace` feature. The default should be `None` which keeps the global setting (which currently defaults to `false` (`Legacy`).
2. Get the `log_namespace` value. You _must_ combine the source and global setting to get the real value. There are helper functions to do this. There are different ways to get this depending on where you need it. See the diff for examples.
3. Update how data is inserted into events to use the `log_namespace`. The existing behavior should stay the same with the `Legacy` namespace. With the new `Vector` namespace, only the decoded data is placed on the event, and _all_ additional data is placed in the event metadata. There are two types of metadata. 
   - "Vector" metadata: standard information inserted by all sources, such as `source_type` and `ingest_timestamp`. These should be inserted with the `LogNamespace::insert_vector_metadata` function.
   - "source" metadata: any additional data not covered already. Those should be inserted with the `LogNamespace::insert_source_metadata` function.
4. Update the schema definition. Many sources don't have one at all yet, so it will need to be added. This will start with the definition of the selected decoder, and then add on any fields that were added in step 3. Any fields added with `LogNamespace::insert_source_metadata` should be added to the definition with `.with_source_metadata`. All "vector" metadata for sources (`ingest_timestamp` and `source_type`) are covered by the single function `.with_standard_vector_source_metadata`.
There are other lower-level methods on the schema definition such as `.with_field`, `.with_metadata_field`, .etc. Those don't look at the `log_namespace` and shouldn't be used unless the previously mentioned methods don't work for some reason.
5. Adjust / add tests as needed. 

When inserting new data, keep in mind that the `Vector` namespace no longer uses the "global log schema", and doesn't use any source configurable key names anymore. It will always use hardcoded keys (since we can now guarantee they won't collide with anything else). When using the `LogNamespace::with_source_metadata` or `schema::Definition::with_source_metadata` functions, you will provide both the `Legacy` and `Vector` name for the field (which in many cases is the same thing).
When choosing the new hard-coded field name, just pick what you think is appropriate and mention it in the PR so the team can discuss / come to a consensus on them.

### Manual testing
If you need to manually test/debug this for any reason, the following config will be helpful. This will give you the value of both the event and metadata as well as the type definitions set for each of them. It may be useful to run this and vary the following settings to see how the output changes
- `schema.enabled` to turn on/off schema definitions (you will still get a definition with it off, it will just be more generic)
- `log_namespace` to switch between `Legacy` (false) and `Vector` (true) namespaces
- `decoding.codec` to see how the codec changes the event structure / type definition.

```toml
schema.enabled = true
schema.log_namespace = true

[sources.input]
type = "demo_logs"
format = "json"
decoding.codec = "bytes"

[transforms.remap]
type = "remap"
inputs = ["input"]
source = """
. = {
   "event": .,
   "event_type": type_def(.),
   "metadata_type": type_def(get_metadata_field(.)),
   "metadata": get_metadata_field(.)
}
"""


[sinks.console]
type = "console"
inputs = ["remap"]
encoding.codec = "json"
```

Example output with the above config
```json
{
  "event": "{\"host\":\"247.229.206.126\",\"user-identifier\":\"benefritz\",\"datetime\":\"26/Jul/2022:15:05:40\",\"method\":\"OPTION\",\"request\":\"/secret-info/open-sesame\",\"protocol\":\"HTTP/1.0\",\"status\":\"503\",\"bytes\":21897,\"referer\":\"https://make.com/user/booperbot124\"}",
  "event_type": {
    "bytes": true
  },
  "metadata": {
    "vector": {
      "ingest_timestamp": "2022-07-26T19:05:40.257408055Z",
      "source_type": "demo_logs"
    }
  },
  "metadata_type": {
    "object": {
      "vector": {
        "object": {
          "ingest_timestamp": {
            "timestamp": true
          },
          "source_type": {
            "bytes": true
          }
        }
      }
    }
  }
}
```

### CODE REVIEW

File: src/sources/demo_logs.rs [hunk 1]
```diff
@@ -242,25 +260,33 @@ impl_generate_config_from_default!(DemoLogsConfig);
...
+        // There is a global and per-source `log_namespace` config. The source config overrides the global setting,
+        // and is merged here.
+        let log_namespace = global_log_namespace.merge(self.log_namespace);
+
+        let schema_definition = self
+            .decoding
+            .schema_definition(log_namespace)
+            .with_standard_vector_source_metadata();
+
+        vec![Output::default(self.decoding.output_type()).with_schema_definition(schema_definition)]
```
Comment (neuronull) [inline src/sources/demo_logs.rs:289]: Could be off here but- wouldn't this implementation of `outputs` be very similar in the other sources and if so, would it make sense to extract that to reduce duplicated code ?

File: src/sources/demo_logs.rs [hunk 2]
```diff
@@ -204,9 +213,18 @@ async fn demo_logs_source(
...
+                        log_namespace.insert_vector_metadata(
+                            log,
+                            log_schema().source_type_key(),
+                            "source_type",
+                            "demo_logs",
+                        );
+                        log_namespace.insert_vector_metadata(
+                            log,
+                            log_schema().timestamp_key(),
+                            "ingest_timestamp",
```
Comment (spencergilbert) [inline src/sources/demo_logs.rs:225]: To clarify, are we moving to set `ingest_timestamp` on every source (regardless of the type of namespace)? That's what it looks like, just wanted to confirm.

**Review by spencergilbert — APPROVED** [review]

File: src/sources/demo_logs.rs (hunk 2 above)
Comment (fuchsnj) [reply src/sources/demo_logs.rs:225]: no, sources with the `Legacy` namespace will use `log_schema().timestamp_key()`, and the `Vector` namespace will use `ingest_timestamp`. These migrations shouldn't change anything when using the `Legacy` namespace.

File: src/sources/demo_logs.rs (hunk 1 above)
Comment (fuchsnj) [reply src/sources/demo_logs.rs:289]: Many sources will contain several calls to `.with_source_metadata(..)` which will be unique to that source. This just happens to be a really simple source. (See the `datadog_agent` which is more complicated).

Other than that, there will be a few lines duplicated, but I'm not seeing a great way to deduplicate that without potentially making it harder to read / configure.

File: src/sources/demo_logs.rs (hunk 2 above)
Comment (spencergilbert) [reply src/sources/demo_logs.rs:225]: Bleh, I was still looking for where we switched between `legacy` and `vector` - forgetting the function takes both paths as args. Sorry late night confusion 😅

**Review by neuronull — APPROVED** [review]

**Pull request merged**